src/pykeen/triples/statement_factory.py

```diff
@@ -0,0 +1,600 @@
+            self.relation_to_inverse = None
+
+    @staticmethod
+    def _check_already_inverted_relations(relations: Iterable[str]) -> bool:
+        for relation in relations:
+            if relation.endswith(INVERSE_SUFFIX):
+                # We can terminate the search after finding the first inverse occurrence
+                return True
+
+        return False
+
+    def create_data_object(self):
+        """Create PyTorch Geometric data object."""
+
+        try:
+            from torch_geometric.data import Data
+        except ImportError as err:
+            raise ImportError("Requires `torch_geometric` to be installed.") from err
+
+        node_features = self._create_node_feature_tensor()
+        edge_index = self._create_edge_index()
+        qualifier_index = self._create_qualifier_index()
+        edge_type = self._create_edge_type()
+
+        entities = torch.as_tensor(
+            data=sorted(set(self.entity_to_id.values())),  # exclude padding entity
+            dtype=torch.long,
+        )
+
+        return Data(
+            x=node_features,
+            edge_index=edge_index,
+            qualifier_index=qualifier_index,
+            edge_type=edge_type,
+            entities=entities,
+        )
+
+    def _compose_qualifier_batch(self, keys, mapping) -> torch.LongTensor:
+        # batch.shape:
+        batch_size = len(keys)
+
+        # Lookup IDs of qualifiers
+        qualifier_ids = [mapping.get(tuple(k.tolist())) for k in keys]
+
+        # Determine maximum length (for dynamic padding)
+        max_len = max(map(len, qualifier_ids))
+
+        # bound maximum length for guaranteed max memory consumption
+        # TODO: num_qualifier_pairs = max_num_qualifier_pairs ?
+        max_len = min(max_len, self.max_num_qualifier_pairs)
+
+        # Allocate result
+        result = torch.full(size=(batch_size, max_len), fill_value=-1, dtype=torch.long)
+
+        # Retrieve qualifiers
+        for i, this_qualifier_ids in enumerate(qualifier_ids):
+            # limit number of qualifiers
+            # TODO: shuffle?
+            this_qualifier_ids = this_qualifier_ids[:max_len]
+            result[i, : len(this_qualifier_ids)] = torch.as_tensor(data=this_qualifier_ids, dtype=torch.long)
+
+        return result
+
+    def _create_qualifier_index(self):
+        """Create a COO matrix of shape (3, num_qualifiers). Only non-zero (non-padded) qualifiers are retained.
+
+        row0: qualifying relations
+        row1: qualifying entities
+        row2: index row which connects a pair (qual_r, qual_e) to a statement index k
+        :return: shape: (3, num_qualifiers)
+        """
+        if self.max_num_qualifier_pairs is None or self.max_num_qualifier_pairs == 0:
+            return None
+
+        qual_relations, qual_entities, qual_k = [], [], []
+
+        # It is assumed that statements are already padded
+        for triple_id, statement in enumerate(self.mapped_statements):
```

Comment:
since mapped_statements is already a torch tensor, we could use torch methods here instead of going to numpy/plain Python and back to torch again.